models.cpp

```diff
@@ -596,41 +610,50 @@ std::valarray<double> Models::ThreeStepAlternative::right_hand_side(const std::v
 
 // class to hold integration hyperparameters
 Models::explEulerParameters::explEulerParameters(const double startTimeValue,
-                      const double endTimeValue,
+                      const std::vector<double> outputTimesValue,
                       const std::valarray<double> initialConditionValues)
 {
   startTime = startTimeValue;
-  endTime = endTimeValue;
+  outputTimes = outputTimesValue;
   initialCondition = initialConditionValues;
 }
 
 
 
 // ODE solver
-std::valarray<double> Models::integrate_ode_explicit_euler(const Models::explEulerParameters solverParameters,
-                                                   const Models::ModelsBase& model,
-                                                   const Models::ParametersBase& modelParameters)
+std::vector<std::valarray<double>> Models::integrate_ode_explicit_euler(const Models::explEulerParameters solverParameters,
+                                                           const Models::ModelsBase& model,
+                                                           const Models::ParametersBase& modelParameters,
+                                                           const double time_step)
 {
+  const std::vector<double> evalTimes = solverParameters.outputTimes;
+
+    std::vector<std::valarray<double>> solution(evalTimes.size());
   std::valarray<double> x = solverParameters.initialCondition;
-  double time_step = 1e-5;
 
   double time = solverParameters.startTime;
-  while (time < solverParameters.endTime)
+  double dt = time_step;
+  for (unsigned int i = 0; i < evalTimes.size(); i++)
   {
-    // advance to next time
-    if (time + time_step > solverParameters.endTime)
+    while (time < evalTimes[i])
     {
-      time_step = solverParameters.endTime - time;
-      time = solverParameters.endTime;
+      //double endTime = evalTimes[i];
+      // advance to next time
+      if (time + dt > evalTimes[i])
+      {
+        dt = evalTimes[i] - time;
```

Comment:
You modify the time step here in order to get to the preset evaluation times. But do you ever recover it to its previous value for the following time steps?